File: src/pgo/model/mpcal/ModularPlusCalBlock.java
package pgo.model.mpcal;

import pgo.model.pcal.PlusCalMacro;
import pgo.model.pcal.PlusCalProcedure;
import pgo.model.pcal.PlusCalProcesses;
import pgo.model.pcal.PlusCalVariableDeclaration;
import pgo.model.tla.TLAUnit;
import pgo.parser.Located;
import pgo.util.SourceLocation;

import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class ModularPlusCalBlock extends ModularPlusCalNode {
	private final Located<String> name;
	private final List<PlusCalVariableDeclaration> variables;
	private final List<ModularPlusCalMappingMacro> mappingMacros;
	private final List<ModularPlusCalArchetype> archetypes;
	private final List<PlusCalMacro> macros;
	private final List<PlusCalProcedure> procedures;
	private final List<TLAUnit> units;
	private final List<ModularPlusCalInstance> instances;
	private final PlusCalProcesses processes;

	public ModularPlusCalBlock(SourceLocation location, Located<String> name,
	                           List<PlusCalVariableDeclaration> variables,
	                           List<TLAUnit> units, List<ModularPlusCalMappingMacro> mappingMacros,
	                           List<ModularPlusCalArchetype> archetypes, List<PlusCalMacro> macros,
	                           List<PlusCalProcedure> procedures, List<ModularPlusCalInstance> instances,
	                           PlusCalProcesses processes) {
		super(location);
		this.name = name;
		this.variables = variables;
		this.units = units;
		this.mappingMacros = mappingMacros;
		this.archetypes = archetypes;
		this.macros = macros;
		this.procedures = procedures;
		this.instances = instances;
		this.processes = processes;
	}

	@Override
	public ModularPlusCalBlock copy() {
		return new ModularPlusCalBlock(
				getLocation(),
				name,
				variables.stream().map(PlusCalVariableDeclaration::copy).collect(Collectors.toList()),
				units.stream().map(TLAUnit::copy).collect(Collectors.toList()),
				mappingMacros.stream().map(ModularPlusCalMappingMacro::copy).collect(Collectors.toList()),
				archetypes.stream().map(ModularPlusCalArchetype::copy).collect(Collectors.toList()),
				macros.stream().map(PlusCalMacro::copy).collect(Collectors.toList()),
				procedures.stream().map(PlusCalProcedure::copy).collect(Collectors.toList()),
				instances.stream().map(ModularPlusCalInstance::copy).collect(Collectors.toList()),
				processes.copy()
		);
	}

	@Override
	public int hashCode() {
		return Objects.hash(name, variables, units, mappingMacros, archetypes, macros, procedures, instances, processes);
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null || obj.getClass() != getClass()) {
			return false;
		}
		ModularPlusCalBlock that = (ModularPlusCalBlock) obj;
		return name.getValue().equals(that.name.getValue()) &&
				Objects.equals(variables, that.variables) &&
				Objects.equals(units, that.units) &&
				Objects.equals(mappingMacros, that.mappingMacros) &&
				Objects.equals(archetypes, that.archetypes) &&
				Objects.equals(macros, that.macros) &&
				Objects.equals(procedures, that.procedures) &&
				Objects.equals(instances, that.instances) &&
				processes.equals(that.processes);
	}

	@Override
	public <T, E extends Throwable> T accept(ModularPlusCalNodeVisitor<T, E> v) throws E {
		return v.visit(this);
	}

	public Located<String> getName() {
		return name;
	}

	public List<PlusCalVariableDeclaration> getVariables() {
		return Collections.unmodifiableList(variables);
	}

	public List<TLAUnit> getUnits() {
		return Collections.unmodifiableList(units);
	}

	public List<ModularPlusCalMappingMacro> getMappingMacros() {
		return Collections.unmodifiableList(mappingMacros);
	}

	public List<ModularPlusCalArchetype> getArchetypes() {
		return Collections.unmodifiableList(archetypes);
	}

	public List<PlusCalMacro> getMacros() {
		return Collections.unmodifiableList(macros);
	}

	public List<PlusCalProcedure> getProcedures() {
		return Collections.unmodifiableList(procedures);
	}

	public List<ModularPlusCalInstance> getInstances() {
		return Collections.unmodifiableList(instances);
	}

	public PlusCalProcesses getProcesses() {
		return processes;
	}
}

File: src/pgo/model/mpcal/ModularPlusCalBlock.scala
package pgo.model.mpcal

import pgo.model.pcal._
import pgo.model.tla.{TLAFunctionDefinition, TLAIdentifier, TLAOperatorDefinition, TLAUnit, TLAUtils}
import pgo.trans.intermediate.DefinitionRegistry
import pgo.util.SourceLocation

import scala.jdk.CollectionConverters._

object ModularPlusCalBlock {
  def from(plusCalAlgorithm: PlusCalAlgorithm) =
    new ModularPlusCalBlock(
      plusCalAlgorithm.getLocation, plusCalAlgorithm.getName, plusCalAlgorithm.getUnits.asScala.toList,
      plusCalAlgorithm.getMacros.asScala.toList, plusCalAlgorithm.getProcedures.asScala.toList,
      Nil, Nil, plusCalAlgorithm.getVariables.asScala.toList, Nil, plusCalAlgorithm.getProcesses)
}

final class ModularPlusCalBlock(loc: SourceLocation, val name: TLAIdentifier, val units: List[TLAUnit],
                          val macros: List[PlusCalMacro], val procedures: List[PlusCalProcedure],
                          val mappingMacros: List[ModularPlusCalMappingMacro], val archetypes: List[ModularPlusCalArchetype],
                          val variables: List[PlusCalVariableDeclaration], val instances: List[ModularPlusCalInstance],
                          val processes: PlusCalProcesses) extends ModularPlusCalNode(loc) {
  def this(loc: SourceLocation, name: TLAIdentifier, units: java.util.List[TLAUnit], macros: java.util.List[PlusCalMacro],
           procedures: java.util.List[PlusCalProcedure], mappingMacros: java.util.List[ModularPlusCalMappingMacro],
           archetypes: java.util.List[ModularPlusCalArchetype], variables: java.util.List[PlusCalVariableDeclaration],
           instances: java.util.List[ModularPlusCalInstance], processes: PlusCalProcesses) =
    this(loc, name, units.asScala.toList, macros.asScala.toList, procedures.asScala.toList,
      mappingMacros.asScala.toList, archetypes.asScala.toList, variables.asScala.toList,
      instances.asScala.toList, processes)

  override def copy: ModularPlusCalBlock =
    new ModularPlusCalBlock(
      getLocation, name.copy, units.map(_.copy()), macros.map(_.copy()), procedures.map(_.copy()),
      mappingMacros.map(_.copy()), archetypes.map(_.copy()), variables.map(_.copy()),
      instances.map(_.copy()), processes.copy)

  override def accept[T, E <: Throwable](v: ModularPlusCalNodeVisitor[T, E]): T = v.visit(this)

  def accept[T, E <: Throwable](v: ModularPlusCalBlockVisitor[T, E]): T = v.visit(this)

  def getName: TLAIdentifier = name

  def getVariables: java.util.List[PlusCalVariableDeclaration] = variables.asJava

  def getUnits: java.util.List[TLAUnit] = units.asJava

  def getMappingMacros: java.util.List[ModularPlusCalMappingMacro] = mappingMacros.asJava

  def getArchetypes: java.util.List[ModularPlusCalArchetype] = archetypes.asJava

  lazy val instantiatedArchetypes: List[ModularPlusCalArchetype] =
    archetypes.filter(arch => instances.exists(inst => inst.getTarget == arch.getName))
  def getInstantiatedArchetypes: java.util.List[ModularPlusCalArchetype] = instantiatedArchetypes.asJava

  def getMacros: java.util.List[PlusCalMacro] = macros.asJava

  def getProcedures: java.util.List[PlusCalProcedure] = procedures.asJava

  def getInstances: java.util.List[ModularPlusCalInstance] = instances.asJava

  def getProcesses: PlusCalProcesses = processes

  override def equals(other: Any): Boolean = other match {
    case that: ModularPlusCalBlock =>
      name == that.name &&
        units == that.units &&
        macros == that.macros &&
        procedures == that.procedures &&
        mappingMacros == that.mappingMacros &&
        archetypes == that.archetypes &&
        variables == that.variables &&
        instances == that.instances &&
        processes == that.processes
    case _ => false
  }

  override def hashCode(): Int = {
    val state = Seq(name, units, macros, procedures, mappingMacros, archetypes, variables, instances, processes)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}